lib/src/group_changes.dart

```diff
@@ -0,0 +1,307 @@
+/// Store a data object summarizing this subgroup, with the test name,
+/// change in result, set of configurations,
+/// and blamelist intersection and union of the changes in this subgroup.
+/// The object is stored in a multi-level map keyed by blamelist intersection
+/// last commit, blamelist intersection first commit, and set of configurations.
+Map<int, Map<int, Map<String, List<ByConfigSetData>>>> computePageData(
+    List<dynamic> changes, List<String> hashes) {
+  final Map<String, int> hashIndex = Map.fromEntries(
+      Iterable.generate(hashes.length, (i) => MapEntry(hashes[i], i)));
+
+  final resultsForTestAndChange =
+      Map<String, Map<String, List<Map<String, dynamic>>>>();
+
+  for (final Map<String, dynamic> change in changes) {
+    String key;
+    if ('true' == change['matches'] ||
+        change['matches'] is bool && change['matches']) {
+      key =
+          "${change['previous_result']} -> <strong>${change['result']} &#x2714;</strong>";
+    } else {
+      key =
+          "${change['previous_result']} -> <strong>${change['result']}</strong>  (expected ${change['expected']})";
+    }
+    resultsForTestAndChange
+        .putIfAbsent(
+            change['name'], () => Map<String, List<Map<String, dynamic>>>())
+        .putIfAbsent(key, () => <Map<String, dynamic>>[])
+        .add(change);
+  }
+
+  configSets = Map<String, List<String>>();
+  final byBlamelist = Map<int, Map<int, Map<String, List<ByConfigSetData>>>>();
+
+  for (final test in resultsForTestAndChange.keys) {
+    for (final results in resultsForTestAndChange[test].keys) {
+      var changes = resultsForTestAndChange[test][results];
+      while (changes.isNotEmpty) {
+        var before = MinMax();
+        for (var change in changes) {
+          before.add(hashIndex[change["previous_commit_hash"]]);
+        }
+        // Sort changes by before, take all where after < first before, repeat
+        var firstSection = changes
+            .where((change) => hashIndex[change["commit_hash"]] < before.min)
+            .toList();
+        changes = changes
+            .where((change) => hashIndex[change["commit_hash"]] >= before.min)
+            .toList();
+        final configs = <String>[];
+        before = MinMax();
+        final after = MinMax();
+        for (var change in firstSection) {
+          configs.add(change["configuration"]);
+          before.add(hashIndex[change["previous_commit_hash"]]);
+          after.add(hashIndex[change["commit_hash"]]);
+        }
+        configs.sort();
+        final formattedSet = configs.join(",<br>");
+        configSets[formattedSet] = configs;
+        final summaryData =
+            ByConfigSetData(formattedSet, test, results, before, after);
+        byBlamelist
+            .putIfAbsent(summaryData.after.max,
+                () => Map<int, Map<String, List<ByConfigSetData>>>())
+            .putIfAbsent(summaryData.before.min,
```

Comment:
And summaryData.before.min is the latest commit that was seen without the change.